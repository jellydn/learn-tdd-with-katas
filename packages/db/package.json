{
  "name": "db",
  "version": "1.0.0",
  "license": "MIT",
  "main": "./src/index.ts",
  "types": "./src/index.ts",
  "exports": {
    ".": "./dist/index.js"
  },
  "scripts": {
    "lib": "tsup",
    "typeorm": "typeorm-ts-node-commonjs"
  },
  "dependencies": {
    "pg": "8.12.0",
    "reflect-metadata": "0.2.2",
    "typeorm": "0.3.20"
  },
  "devDependencies": {
    "@swc/core": "1.6.13",
    "@types/node": "20.14.10",
    "ts-node": "10.9.2",
    "tsconfig": "workspace:*",
    "tsup": "8.1.0",
    "typescript": "5.5.3"
  },
  "tsup": {
    "entry": [
      "src/**/*.ts"
    ],
    "splitting": false,
    "sourcemap": true,
    "clean": true,
    "dts": true,
    "format": [
      "cjs"
    ]
  }
}
