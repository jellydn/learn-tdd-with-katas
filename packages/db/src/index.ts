import { AppDataSource } from "./data-source";
import { User } from "./entity/User";

export function initDbConnection() {
	return AppDataSource.initialize();
}

// add new user
export async function addUser(user: Partial<User>) {
	const userRepo = AppDataSource.getRepository(User);
	const newUser = userRepo.create(user);
	return userRepo.save(newUser);
}

// get all users
export async function getAllUsers() {
	const userRepo = AppDataSource.getRepository(User);
	return userRepo.find();
}

// get user by id
export async function getUserById(id: number) {
	const userRepo = AppDataSource.getRepository(User);
	return userRepo.findOneOrFail({
		where: {
			id,
		},
	});
}

// update user
export async function updateUser(id: number, user: Partial<User>) {
	const userRepo = AppDataSource.getRepository(User);
	const userToUpdate = await userRepo.findOneOrFail({
		where: {
			id,
		},
	});
	userRepo.merge(userToUpdate, user);
	return userRepo.save(userToUpdate);
}

//soft delete user
export async function deleteUser(id: number) {
	const userRepo = AppDataSource.getRepository(User);
	const userToDelete = await userRepo.findOneOrFail({
		where: {
			id,
		},
	});
	return userRepo.softDelete({
		id: userToDelete.id,
	});
}
