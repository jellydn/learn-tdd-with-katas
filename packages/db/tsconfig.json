{
	"extends": "tsconfig/nodejs.json",
	"compilerOptions": {
		"emitDecoratorMetadata": true,
		"experimentalDecorators": true
	},
	"include": ["src/**/*.ts"]
}
