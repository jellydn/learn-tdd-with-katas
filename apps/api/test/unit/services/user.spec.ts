import * as db from "db";
import { fromPartial } from "@total-typescript/shoehorn";
import { Errors, ServiceBroker } from "moleculer";
import { afterAll, beforeAll, describe, expect, it, vi } from "vitest";

import TestService from "../../../services/user.service";

const { ValidationError } = Errors;

vi.mock("db");

describe("Test 'user' service", () => {
	const broker = new ServiceBroker({ logger: false });
	broker.createService(TestService);

	beforeAll(async () => broker.start());
	afterAll(async () => {
		vi.restoreAllMocks();
		return broker.stop();
	});

	describe("Test 'user.all' action", () => {
		it("should return with all users", async () => {
			vi.spyOn(db, "getAllUsers").mockResolvedValue([
				{
					id: 1,
					firstName: "Dung",
					lastName: "Huynh",
					age: 35,
				},
			]);
			const response = await broker.call("user.all");

			expect(response).toMatchSnapshot();
		});
	});

	describe("Test 'user.profile' action", () => {
		it("should return with profile", async () => {
			vi.spyOn(db, "getUserByUserId").mockResolvedValue(
				fromPartial({
					id: 1,
					age: 35,
					firstName: "Dung",
					lastName: "Huynh",
				})
			);
			const response = await broker.call("user.profile", {
				userId: "1",
			});

			expect(response).toMatchSnapshot();
		});

		it("should reject an ValidationError", async () => {
			expect.assertions(1);
			try {
				await broker.call("user.profile");
			} catch (error: unknown) {
				expect(error).toBeInstanceOf(ValidationError);
			}
		});
	});
});
