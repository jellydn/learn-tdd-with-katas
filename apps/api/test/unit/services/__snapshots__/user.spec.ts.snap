// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test 'user' service > Test 'user.all' action > should return with all users 1`] = `undefined`;

exports[`Test 'user' service > Test 'user.profile' action > should return with profile 1`] = `undefined`;
